### PR TITLE
add suport for exposing ignition service as node port

### PR DESCRIPTION
**Pull request by relyt0925**
Solves:
https://github.com/openshift/hypershift/issues/365

### CODE REVIEW

File: api/v1alpha1/hostedcluster_types.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ type HostedClusterSpec struct {
 // ServicePublishingStrategyMapping defines the service being published and  metadata about the publishing strategy.
 type ServicePublishingStrategyMapping struct {
 	// Service identifies the type of service being published
-	// +kubebuilder:validation:Enum=APIServer;OAuthServer;OIDC;Konnectivity
+	// +kubebuilder:validation:Enum=APIServer;OAuthServer;OIDC;Konnectivity;Ignition
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:104]: this is orthogonal to this PR by as anote for future API refinements I wonder if this validation should live in the ServiceType type definition.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -279,15 +279,37 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// Set Ignition Server endpoint
 	{
+		serviceStrategy := servicePublishingStrategyByType(hcluster, hyperv1.Ignition)
+		if serviceStrategy == nil {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:283]: Although structural schema enum validation should prevent this from ever be the case I wonder in addition to logs should we also set a condition that signal this in a more ux friendly manner?

I wonder should we log the error but not return it here? since this won't get fixed by requeueing unless there's and update to the API which will already emit an update event a reconciliation will happen.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 3]
```diff
@@ -279,15 +279,37 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+				}
 			}
-		}
-		if err == nil && ignitionServerRoute.Spec.Host != "" {
-			hcluster.Status.IgnitionEndpoint = ignitionServerRoute.Spec.Host
+			if err == nil && ignitionServerRoute.Spec.Host != "" {
+				hcluster.Status.IgnitionEndpoint = ignitionServerRoute.Spec.Host
+			}
+		case hyperv1.NodePort:
+			if serviceStrategy.NodePort == nil {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:299]: in addition to logs should we also set a condition that signal this in a more ux friendly manner?

I wonder should we log the error but not return it here? since this won't get fixed by requeueing unless there's and update to the API which will already emit an update event a reconciliation will happen.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 4]
```diff
@@ -279,15 +279,37 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+				return ctrl.Result{}, fmt.Errorf("nodeport metadata not specified for ignition service")
+			}
+			ignitionService := ignitionserver.Service(controlPlaneNamespace.GetName())
+			if err := r.Client.Get(ctx, client.ObjectKeyFromObject(ignitionService), ignitionService); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return ctrl.Result{}, fmt.Errorf("failed to get ignition service: %w", err)
+				}
+			}
+			if err == nil && len(ignitionService.Spec.Ports) == 1 && ignitionService.Spec.Ports[0].NodePort > 0 {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:308]: nit: It'd be nice to move `len(ignitionService.Spec.Ports) == 1 && ignitionService.Spec.Ports[0].NodePort > 0` into a unit tested function e.g `ServiceNodePortAvailable` or so.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 5]
```diff
@@ -1019,6 +1041,44 @@ func (r *HostedClusterReconciler) reconcileControlPlaneOperator(ctx context.Cont
 	return nil
 }
 
+func servicePublishingStrategyByType(hcp *hyperv1.HostedCluster, svcType hyperv1.ServiceType) *hyperv1.ServicePublishingStrategy {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1044]: nit: it'd be nice to have a simple unit test for this.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 6]
```diff
@@ -1019,6 +1041,44 @@ func (r *HostedClusterReconciler) reconcileControlPlaneOperator(ctx context.Cont
...
+	var portSpec corev1.ServicePort
+	if len(svc.Spec.Ports) > 0 {
+		portSpec = svc.Spec.Ports[0]
+	} else {
+		svc.Spec.Ports = []corev1.ServicePort{portSpec}
+	}
+	portSpec.Port = int32(443)
+	portSpec.Name = "https"
+	portSpec.Protocol = corev1.ProtocolTCP
+	portSpec.TargetPort = intstr.FromInt(9090)
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1066]: why is this undoing `portSpec.TargetPort = intstr.FromString("https")`?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 7]
```diff
@@ -1019,6 +1041,44 @@ func (r *HostedClusterReconciler) reconcileControlPlaneOperator(ctx context.Cont
...
+func servicePublishingStrategyByType(hcp *hyperv1.HostedCluster, svcType hyperv1.ServiceType) *hyperv1.ServicePublishingStrategy {
+	for _, mapping := range hcp.Spec.Services {
+		if mapping.Service == svcType {
+			return &mapping.ServicePublishingStrategy
+		}
+	}
+	return nil
+}
+
+func reconcileIgnitionServerService(svc *corev1.Service, strategy *hyperv1.ServicePublishingStrategy) error {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1053]: nit: it'd be nice to have this unit tested.

**Comment by enxebre** [comment]
I'd be nice to include an e2e test for this.

**Comment by relyt0925** [comment]
We are going to do a full IBM E2E test once these downstream ones are in and out it as a gate

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 2 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:283]: I think we need to return error as it doesn't make sense to continue execution if this has failed. However I could also log if that helps.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:299]: I think we need to return error as it doesn't make sense to continue execution if this has failed. However I could also log if that helps.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 5 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1044]: fixing....

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 7 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1053]: fixing....

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 6 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1066]: this failed on reconciliation updates. I believe it needs to be the port number.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 4 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:308]: fixing....

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 2 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:283]: Yeh we need to break execution but we don't necessarily need to return an error.
If we return error the controller runtime implementation will automatically requeue the event again and again. In this case we know that this particular problem will never be fixed by requeueing because the problem is a bad input.
So I wonder if instead of returning the error we want to just to log it and return nil.
Basically this:
https://github.com/openshift/hypershift/blob/b7566016fc3a287435f32d625abdf447e458e301/hypershift-operator/controllers/nodepool/nodepool_controller.go#L243-L246

Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:290]: OHHHHH I see what you mean

Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:290]: This is neat sorry no that makes sense......

Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:290]: I will change to this I agree pointless to auto requeue

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:306]: fixing agreed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 5 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1060]: fixed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 7 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1069]: fixed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 4 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: fixed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:309]: fixed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 2 above)
Comment (relyt0925) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:290]: fixed

**Comment by enxebre** [comment]
Thanks @relyt0925! I appreciate the unit tests.
I'll defer the second tag for a second pair of eyes cc @ironcladlou @sjenning @csrwng 
/approve
